make the ramdisk feature functional

lemin9538 opened the pull request:
map vm region
get ramdisk from fdt
copy fdt before setup vm
copy kernel before start vm
add a script that can make ramdisk

lemin9538 commented:
@l3ifeng I update your patch, please have a review, and could you help to test it :) ? If the ramdisk feature is ok, then I can prepare to  the implement the natvie VM reset feature.

l3ifeng commented:
@lemin9538 I will test it soon.

l3ifeng commented on core/mm.c:
```diff
@@ -176,14 +177,22 @@ static void map_os_memory(void)
 		if (type == MEMORY_REGION_TYPE_DMA)
 			flags |= VM_IO;
 		else
 			flags |= VM_NORMAL;
```
Is it appropriate to treat `VM` and `ramdisk` regions as `VM_NORMAL`?

lemin9538 replied:
yes, need add MEMORY_REGION_TYPE_RAMDISK.

l3ifeng commented on core/mm.c:
```diff
@@ -176,14 +177,22 @@ static void map_os_memory(void)
-				(type != MEMORY_REGION_TYPE_DMA))
+		    (type != MEMORY_REGION_TYPE_DMA) &&
+		    /*
+		     * Map the memory in VM region, because we may touch the
```
`VM` and `ramdisk`

l3ifeng commented on core/mm.c:
```diff
@@ -176,14 +177,22 @@ static void map_os_memory(void)
+		     * Map the memory in VM region, because we may touch the
+		     * kernel and fdt.
+		     */
+		    (type != MEMORY_REGION_TYPE_VM))
```
`MEMORY_REGION_TYPE_VM` and `MEMORY_REGION_TYPE_RAMDISK`

l3ifeng commented on core/mm.c:
```diff
@@ -176,14 +177,22 @@ static void map_os_memory(void)
 			flags |= VM_NORMAL;
 
+		if (type == MEMORY_REGION_TYPE_RAMDISK)
+			flags |= VM_RO;
```
It's weird that this flag will cause `sync error in EL2`.

lemin9538 replied:
yes, there is a mistake, it should be : flags |= VM_NORMAL | VM_RO.

l3ifeng replied:
It is `if` instead of `else if`, which means it is already `VM_NORMAL | VM_RO`.

l3ifeng replied:
It seems that Minos cannot access ramdisk during `ramdisk_init` if `flags |= VM_RO`.

lemin9538 replied:
After thinking about it, there will be problems with mapping here. The memory of the ramdisk belongs to the kernel, and all kernel memory has been mapped when it is initialized. The logic of this piece needs to be changed。

l3ifeng replied:
You use the `S1_` prefix macro in `arch_host_tt_description` function. Should the memory used by Minos belong to stage 2, right?

https://developer.arm.com/architectures/learn-the-architecture/aarch64-memory-model/combining-stage-1-and-stage-2-attributes

lemin9538 replied:
stage2 is for VM. Minos is stage1. ramdisk memory region has already been mapped in arch/aarch64/core/mem_map.S, there should some issue when change the mapping for those memory which already mapped.

l3ifeng replied:
The location of the ramdisk does not belong to the kernel, it is not mapped. I have already tried it.

lemin9538 replied:
ok got it, you load the ramdisk into the memory region which do not belongs to kernel.

l3ifeng replied:
Yes, that is why we need the `ramdisk` region.

lemin9538 replied:
which address do you load ? create_host_mapping will map the memory as 2M align?

l3ifeng replied:
I mapped it to 0x4a000000.

lemin9538 replied:
list_for_each_entry(region, &mem_list, list) {
                type = memory_region_type(region);
+                /*
+                 * Map the memory in VM region, because we may touch the
+                 * kernel and fdt.
+                 */
                if ((type != MEMORY_REGION_TYPE_NORMAL) &&
-                               (type != MEMORY_REGION_TYPE_DMA))
+                   (type != MEMORY_REGION_TYPE_DMA) &&
+                   (type != MEMORY_REGION_TYPE_VM) &&
+                   (type != MEMORY_REGION_TYPE_RAMDISK));
                        continue;

please try it again.

l3ifeng commented on core/mm.c:
```diff
@@ -175,15 +176,24 @@ static void map_os_memory(void)
 		else
 			flags |= VM_NORMAL;
 
+		if (type == MEMORY_REGION_TYPE_RAMDISK)
+			flags |= VM_RO;
```
There are still memory access errors, but this time the error occurred in `ramdisk_read`. Just remove this paragraph and it will work. I guess the `VM` area is also read-only?

lemin9538 replied:
Ok, could you help debug this issue. you can modify this patch directly if you need.

l3ifeng replied:
Surely

l3ifeng replied:
Ah, I found it. Change all `|=` to `=`. The `flags` are outside the `list_for_each_entry`, causing `VM_RO` to always exist.

lemin9538 replied:
good.

l3ifeng replied:
I will clean and retest this patch. Thank you.

l3ifeng commented:
@lemin9538 I fixed, cleaned and retested this patch. I rewrote the commit message. I don't know how to submit to this PR. You can get it [here](https://github.com/l3ifeng/minos/commits/ramdisk).

lemin9538 commented:
> @lemin9538 I fixed, cleaned and retested this patch. I rewrote the commit message. I don't know how to submit to this PR. You can get it [here](https://github.com/l3ifeng/minos/commits/ramdisk).

Ok, I can push it.

lemin9538 commented:
LGTM
